Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -0,0 +1,356 @@
+            case .https:
+                return bootstrap.connect(host: self.key.host, port: self.key.port)
+            case .https_unix:
+                return bootstrap.connect(unixDomainSocketPath: self.key.unixPath)
+            case .http, .http_unix, .unix:
+                preconditionFailure("Unexpected scheme")
+            }
+        }.flatMap { channel -> EventLoopFuture<(Channel, String?)> in
+            let tlsEventHandler = try! channel.pipeline.syncOperations.handler(type: TLSEventsHandler.self)
+            return tlsEventHandler.tlsEstablishedFuture.flatMap { negotiated in
+                channel.pipeline.removeHandler(tlsEventHandler).map { (channel, negotiated) }
+            }
+        }
+
+        #if canImport(Network)
+            // If NIOTransportSecurity is used, we want to map NWErrors into NWPOsixErrors or NWTLSError.
+            channelFuture = channelFuture.flatMapErrorThrowing { error in
+                throw HTTPClient.NWErrorHandler.translateError(error)
+            }
+        #endif
+
+        return channelFuture
+    }
+
+    private func makeTLSBootstrap(eventLoop: EventLoop, logger: Logger)
+        -> EventLoopFuture<NIOClientTCPBootstrapProtocol> {
+        // since we can support h2, we need to advertise this in alpn
+        var tlsConfig = self.tlsConfiguration
+        tlsConfig.applicationProtocols = ["http/1.1" /* , "h2" */ ]
+
+        #if canImport(Network)
+            if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *), let tsBootstrap = NIOTSConnectionBootstrap(validatingGroup: eventLoop) {
+                // create NIOClientTCPBootstrap with NIOTS TLS provider
+                let bootstrapFuture = tlsConfig.getNWProtocolTLSOptions(on: eventLoop).map {
+                    options -> NIOClientTCPBootstrapProtocol in
+
+                    tsBootstrap
+                        .addTimeoutIfNeeded(self.clientConfiguration.timeout)
+                        .tlsOptions(options)
+                        .channelInitializer { channel in
+                            do {
+                                try channel.pipeline.syncOperations.addHandler(HTTPClient.NWErrorHandler())
+                                // we don't need to set a TLS deadline for NIOTS connections, since the
+                                // TLS handshake is part of the TS connection bootstrap. If the TLS
+                                // handshake times out the complete connection creation will be failed.
+                                try channel.pipeline.syncOperations.addHandler(TLSEventsHandler(deadline: nil))
+                                return channel.eventLoop.makeSucceededFuture(())
+                            } catch {
+                                return channel.eventLoop.makeFailedFuture(error)
+                            }
+                        } as NIOClientTCPBootstrapProtocol
+                }
+                return bootstrapFuture
+            }
+        #endif
```

Comment:
This section is kind of rough to read, could it be refactored a little to have less indentation?